### ISSUE TITLE
New healthcheck for flatmap drawing from just

### ISSUE THREAD

**Issue by Zac-HD** [HypothesisWorks/hypothesis#1604]
There are two ways to draw a value from some strategy `s` and apply a function `f` to the value:

- `s.map(f)` is the right way to do it.
- `s.flatmap(lambda x: just(f(x)))` is... not.  It *works*, but it's slow and doesn't shrink so well.

However, the flatmap form does crop up in the wild - notably causing performance problems in #114, and also in [`hs-dbus=strategies`](https://github.com/stratis-storage/hs-dbus-signature/pull/16).  

We can help users avoid this by adding `HealthCheck.just_in_flatmap` with a message recommending use of `.map`, to be raised if the function passed to `.flatmap` returns an instance of `JustStrategy`.

**Comment by DRMacIver** [comment]
I don't think this can be made to work correctly. e.g. `s.flatmap(lambda x: some_strategy if x else just(42))` is perfectly valid, so you need something more than checking if the result is a `just`.

On top of that, I'm not convinced it would be useful. `flatmap` shrinks identically to the equivalent map these days, and it shouldn't be notably slow except in the sense of involving slightly more function calls.
